dask/layers.py

```diff
@@ -1239,3 +1262,341 @@ def __repr__(self):
         return "DataFrameIOLayer<name='{}', n_parts={}, columns={}>".format(
             self.name, len(self.inputs), self.columns
         )
+
+
+class DataFrameTreeReduction(DataFrameLayer):
+    """DataFrame Tree-Reduction Layer
+
+    Parameters
+    ----------
+    name : str
+        Name to use for the constructed layer.
+    name_input : str
+        Name of the input layer that is being reduced.
+    npartitions_input : str
+        Number of partitions in the input layer.
+    concat_func : callable
+        Function used by each tree node to reduce a list inputs into
+        a single output value. This function must accept only a list
+        as its first positional argument.
+    tree_node_func : callable
+        Function used on the output of ``concat_func`` in each tree
+        node. This function must accept the output of ``concat_func``
+        as its first positional argument. Any other input arguments
+        must be definded by in ``tree_node_kwargs``.
+    tree_node_args : dict, optional
+        Dictionary of key-word arguments to include in every call to
+        the function specified by ``tree_node_func``. Note that the
+        first positional argument will always correspond to a list of
+        inputs from the input nodex (and these arguments will always
+        come after).
```

Comment:
I'm actually very interested to establish if `functools.partial` is really an option here.  After looking at [the documentation](https://docs.python.org/3/library/functools.html#partial-objects), my understanding was that we can only push positional arguments of the function into a partial object if the pre-defined values come first. This means `partial` will not work if the first argument of `tree_node_func` is the input data.  Maybe I am missunderstanding the behavior?

Reply:
```python
In [1]: from functools import partial

In [2]: def foo(*args, **kwargs):
   ...:     print(args, kwargs)
   ...: 

In [3]: partial(foo, k1=1, k2=2)("a", "b")
('a', 'b') {'k1': 1, 'k2': 2}
```
Seems like you can bind kwargs first and still apply positional arguments later.

Reply:
Is this demonstrating anything about binding positional args? I know we can bind kwargs into a partial object, but the `tree_node_args` are meant to be positional arguments only.

Reply:
The behavior in your example still seems problematic to me. This would suggest that `foo` must expect the input data as the first positional argument **after** the arguments specified with `partial`. I get the impression that it is much more likely that `foo` would expect the data input as the very first positional argument.

Reply:
>Oh, I'm confused since the docstring says it's a "Dictionary of key-word arguments".

Oops - My bad (again).

Reply:
Okay - I was actually the one who was confused here. I do still think positional args are a pain in `partial`, but this was indeed meant for key-word  arguments anyway.

I have no problem requiring the caller to push all kwargs into the function into a `partial` object. We do this automatically in `__dask_distributed_pack__` anyway. The only reason to include these parameters is for a bit of code clarity in places where a DataFrameTreeReduction layer is defined. I can certainly remove them :)

Reply:
Yeah, I don't think it matters too much, I just think it's somewhat nice to simplify the interface so there are fewer parameters.